pilot/pkg/status/state.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/ghodss/yaml"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+
+	"istio.io/pkg/log"
+)
+
+var scope = log.RegisterScope("status",
+	"component for writing distribution status to istiio CRDs", 0)
+
+type Progress struct {
+	AckedInstances int
+	TotalInstances int
+}
+
+func (p *Progress) PlusEquals(p2 Progress) {
+	p.TotalInstances += p2.TotalInstances
+	p.AckedInstances += p2.AckedInstances
+}
+
+type DistributionController struct {
+	mu               sync.RWMutex
+	CurrentState     map[Resource]map[string]Progress
+	ObservationTime  map[string]time.Time
+	UpdateInterval   time.Duration
+	client           dynamic.Interface
+	clock            clock.Clock
+	knownResources   map[schema.GroupVersionResource]dynamic.NamespaceableResourceInterface
+	currentlyWriting ResourceLock
+	StaleInterval    time.Duration
+	QPS              float32
+	Burst            int
+}
+
+func (c *DistributionController) Start(restConfig *rest.Config, namespace string, stop <-chan struct{}) {
```

Comment:
How about adding an UpdateStatus to the k8s config registry - that already tracks all Istio resources and has pretty much all this code first hand. 

And while doing this - cleanup the Status updater for Ingress ( and possibly for Gateway @howardjohn )

Reply:
With one day left before the branch is cut, this seems an unwise time to begin a refactor of pilot's k8s interfaces... What's wrong with using k8s libraries directly?

Reply:
We just finished removing a massive set of duplications and inconsistent ways to do things in galley.